contributors/design-proposals/daemonset-create-first-rolling-update.md

```diff
@@ -0,0 +1,178 @@
+
+1. Determine how many nodes can have new pods scheduled, according to the
+   number of extra nodes currently running more than one pod and the `maxSurge`
+   parameter.
+2. Determine how many nodes can have their previous-generation pods retired
+   once the replacement pods are scheduled and running.
+3. Instruct the controller-manager to add and delete pods on nodes based on the
+   results of (1) and (2).
+
+This is mechanically almost identical to the current `RollingUpdate` strategy,
+with the key differences being the direct creation of new pods on nodes (the
+`RollingUpdate` strategy only deletes pods up to the `maxUnavailable` threshold
+and then waits for the main reconciler create new pods to replace them).
+
+One minor complication is that the controller-manager's reconciliation loop
+actively tries to enforce the one-pod-per-node invariant, and work is underway
+to fix cases in which the invariant is not held
+[kubernetes/kubernetes#50477](https://github.com/kubernetes/kubernetes/issues/50477).
+However, this invariant will need to be relaxed when the `SurgingRollingUpdate`
+strategy temporarily creates extra pods.
+
+### Alternatives considered
+
+The `maxSurge` parameter could have been added to the existing `RollingUpdate`
+strategy (as is the case for Deployments). However, this would break validation
+backwards compatibility, since the `maxUnavailable` parameter currently
+requires `maxUnavailable` to be > 0, but this is not required with
+`SurgingRollingUpdate`.
+
+There are plans to create a `v1beta2` iteration of the apps API (enabled by
+default in 1.9+), so it may be possible to add the `maxSurge` strategy to
+`RollingUpdate` and deprecate `SurgingRollingUpdate` then, or make the change
+now since the `v1beta2` API is incoming. However, to be conservative it was
+decided to implement this as a separate strategy.
+
+For more background see https://github.com/kubernetes/kubernetes/issues/48841.
+
+### Considerations / questions
+
+1. How are `hostPort`s handled? 
+
+They are not handled as part of this proposal. We can either:
+
+  1. Attempt to determine a mechanism to handle this (unclear, perhaps best
+     left as future work, i.e. prior to GA)
+  2. Note in the documentation that this may cause port conflicts that prevent
+     new pods from scheduling successfully (and therefore updates from
+     completing successfully)
+  3. Add a validation check that rejects the combination of `hostPort` and the
+     `SurgingRollingUpdate` strategy.
+
+2. How will the scheduler handle hostPort collisions?
+
+The pods will not be scheduled. This is also true of Deployments (the new pods
+can get stuck in pending). To verify, create a new Deployment that uses a
+hostPort, has replicas = the number of workers (or use a selector), and set
+`maxUnavailable` to 0. If you attempt to update this Deployment the update will
+get stuck since the scheduler cannot find any nodes upon which to place the
+pods. 
```

Comment:
It's common for a DaemonSet to specify `hostPort` in its pod template, but not for Deployments.

Reply:
Added a note to that effect for full honesty.